### ISSUE TITLE
Waiting for events after they have been fired

### ISSUE THREAD

**Issue by fschuetz04** [luteberget/simcpp#1]
Waiting for an event which is fired afterward works as expected:

```cpp
#include "simcpp.h"

class Waiter : public Process {
public:
  Waiter(std::shared_ptr<Simulation> sim, std::shared_ptr<Event> event)
      : Process(sim), event(event) {}

  virtual bool run() override {
    PT_BEGIN();
    printf("Waiting for event\n");
    PROC_WAIT_FOR(event);
    printf("Event fired\n");
    PT_END();
  }

private:
  std::shared_ptr<Event> event;
};

int main() {
  auto sim = Simulation::create();
  auto event = std::make_shared<Event>(sim);
  sim->start_process<Waiter>(event);
  event->fire();
}
```

However, calling `event->fire()` before starting the `Waiter` process produces a segementation fault, as `listeners` is set to `nullptr` and thus `event->add_handler(...)` (called by `PROC_WAIT_FOR(...)`) no longer works. For me, this concept makes sense -- and it also works in SimPy ;).

It would be possible to check whether the event has been fired via `event->is_processed()`, but doing this every time seems a bit complicated. Instead, I would suggest adding a check to `PROC_WAIT_FOR(...)`.

What do you think? I can also submit a pull request if desired.

**Comment by luteberget** [comment]
Thanks for the report. As you might have guessed, this repository was an experiment that I abandoned after a short time, so I don't remember at this point how everything was supposed to hang together. :)

I made an attempt anyway to fix the specific problem you are describing. Added the check for `event->is_triggered()` to the `PROC_WAIT_FOR` macro.

If you have further suggestions for cleaning up this code I would be happy to hear about it.

**Comment by fschuetz04** [comment]
Thanks for the fix!

Did you still continue with discrete-event simulation, and if so, what frameworks did you find useful?

About cleaning up the code: I guess the whole thing could be improved with C++20 coroutines, but as they are not easy to use as of now, I would not do it. Overall, the code is clean enough in my opinion (though for myself, I split it into .cpp and .h to include it in multiple files).

**Comment by fschuetz04** [comment]
Actually, another problem occured with the fix you commited: If the processes waiting for an event immediately waits for the same event again (or starts another process waiting for the same event), a segmentation fault occurs. `value` should be set before resuming the processes to prevent this.

**Comment by luteberget** [comment]
Having co-routines in the language would make discrete event simulation so much better. I actually gave up on coroutines and wrote a railway simulation in Rust with explicit state machine code instead of coroutines.

You can see the simulation setup here: https://github.com/koengit/trainspotting/blob/master/rolling/src/eventsim/simulation.rs
You can see an example of simulation code here: https://github.com/koengit/trainspotting/blob/master/rolling/src/railway/infrastructure.rs#L85

I'm not claiming this is elegant code, but I found this approach to be easier to wrap my head around than trying to hack together co-routines a language that don't really support it. I'm also not a C++ expert.

I tried again to fix the issue, feel free to correct me again or make a pull request :) Thanks.

**Comment by fschuetz04** [comment]
Yeah, that's why I chose SimPy for my simulation, but now I want to check whether a compiled simulation would bring a major speed-up (compared to JIT compilation with PyPy).

Thanks for the examples! I think I like SimPy more, but I guess then it wouldn't be a fast simulation.

And thanks again for the fix, see my pull request #2 for further changes.
